composer/trainer/dist_strategy.py

```diff
@@ -369,78 +369,141 @@ def sync_hook(*args):
-                    module.reset_parameters()
-                else:
-                    raise ValueError(
-                        f'Object `{obj_name}` does not have a ``param_init_fn`` or a ``reset_parameters`` function. '
-                        'This leaves parameters without initialization. Please add a ``param_init_fn`` or ``reset_parameters`` '
-                        f'to module `{obj_name}`.')
+
+                    # Arbitrarily choose the first module as the source module
+                    first_mod, first_attr = mod_attr_list[0]
+                    source_mod_to_mod_attr[first_mod] = [
+                        (target_mod, first_attr, dest_attr) for target_mod, dest_attr in mod_attr_list[1:]
+                    ]
+
+                # Clean up no longer needed module references for memory safety
+                del tied_pointers
+
+                def _param_init_fn(module: torch.nn.Module) -> None:
+                    # If we do not have any parameters or buffers on meta device managed by this module directly, we do not need to call the parameter init function.
+                    # It is assumed that whatever process moved the parameters off of meta device initialized them.
+                    # We expect this to occur if we have tied weights, as the second module will already have the weights initialized.
+                    is_meta = any(param.is_meta for param in module.parameters(recurse=False)) or any(
+                        buffer.is_meta for buffer in module.buffers(recurse=False))
+                    if not is_meta:
+                        return
+
+                    # Move all parameters and buffers to the current device
+                    module.to_empty(device=f'cuda:{torch.cuda.current_device()}', recurse=False)
+
+                    # Redo weight tying, which will have been broken by the above line that moves parameters off of meta device
+                    if module in source_mod_to_mod_attr:
+                        for target_mod, first_attr, dest_attr in source_mod_to_mod_attr[module]:
+                            setattr(target_mod, dest_attr, getattr(module, first_attr))
+
+                    # Run the specified initialization
+                    if hasattr(obj, 'param_init_fn') and isinstance(obj.param_init_fn, Callable):
+                        obj.param_init_fn(module)
+                    elif hasattr(module, 'reset_parameters') and isinstance(module.reset_parameters, Callable):
+                        module.reset_parameters()
+                    else:
+                        raise ValueError(
+                            f'Object `{obj_name}` does not have a ``param_init_fn`` or a ``reset_parameters`` function. '
+                            'This leaves parameters without initialization. Please add a ``param_init_fn`` or ``reset_parameters`` '
+                            f'to module `{obj_name}`.')
+            else:
```

Comment:
Just to verify
Even tho its showing up as new code, the entire else block is what we previously had 
ie it doesn't need to be checked?

As a meta note: this will change init if using torch2.1+ vs torch < 2.1, right?

Reply:
yes, the entire block in the `else` is copied from before

yes, meta init will have different initialization.

Reply:
> yes, the entire block in the else is copied from before

👍 

> yes, meta init will have different initialization.

😞 
ok thats fine.